spring-kafka/src/main/java/org/springframework/kafka/listener/ConditionalDelegatingBatchErrorHandler.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * An error handler that delegates to different error handlers, depending on the exception
+ * type.
+ *
+ * @author Gary Russell
+ * @since 2.7.4
+ *
+ */
+public class ConditionalDelegatingBatchErrorHandler implements ContainerAwareBatchErrorHandler {
+
+	private final ContainerAwareBatchErrorHandler defaultErrorHandler;
+
+	private final Map<Class<? extends Throwable>, ContainerAwareBatchErrorHandler> delegates = new LinkedHashMap<>();
+
+	/**
+	 * Construct an instance with a default error handler that will be invoked if the
+	 * exception has no matches.
+	 * @param defaultErrorHandler the default error handler.
+	 */
+	public ConditionalDelegatingBatchErrorHandler(ContainerAwareBatchErrorHandler defaultErrorHandler) {
+		Assert.notNull(defaultErrorHandler, "'defaultErrorHandler' cannot be null");
+		this.defaultErrorHandler = defaultErrorHandler;
+	}
+
+	/**
+	 * Set the delegate error handlers; a {@link LinkedHashMap} argument is recommended so
+	 * that the delegates are searched in a known order.
+	 * @param delegates the delegates.
+	 */
+	public void setErrorHandlers(Map<Class<? extends Throwable>, ContainerAwareBatchErrorHandler> delegates) {
+		this.delegates.clear();
+		this.delegates.putAll(delegates);
+	}
+
+	/**
+	 * Add a delegate to the end of the current collection.
+	 * @param throwable the throwable for this handler.
+	 * @param handler the handler.
+	 */
+	public void addDelegate(Class<? extends Throwable> throwable, ContainerAwareBatchErrorHandler handler) {
+		this.delegates.put(throwable, handler);
+	}
+
+	@Override
+	public void handle(Exception thrownException, ConsumerRecords<?, ?> records, Consumer<?, ?> consumer,
+			MessageListenerContainer container) {
+
+		// Never called but, just in case
+		doHandle(thrownException, records, consumer, container, null);
+	}
+
+	@Override
+	public void handle(Exception thrownException, ConsumerRecords<?, ?> records, Consumer<?, ?> consumer,
+			MessageListenerContainer container, Runnable invokeListener) {
+
+		doHandle(thrownException, records, consumer, container, invokeListener);
+	}
+
+	protected void doHandle(Exception thrownException, ConsumerRecords<?, ?> records, Consumer<?, ?> consumer,
+			MessageListenerContainer container, @Nullable Runnable invokeListener) {
+
+		boolean handled = false;
+		Throwable cause = thrownException;
+		if (cause instanceof ListenerExecutionFailedException) {
+			cause = thrownException.getCause();
+		}
+		if (cause != null) {
+			Class<? extends Throwable> causeClass = cause.getClass();
+			for (Entry<Class<? extends Throwable>, ContainerAwareBatchErrorHandler> entry : this.delegates.entrySet()) {
+				if (entry.getKey().equals(causeClass)) {
+					handled = true;
+					entry.getValue().handle(thrownException, records, consumer, container, invokeListener);
+					break;
```

Comment:
Why just `return;` from here is not enough instead of that `handled` state tracking?